### PR TITLE
[5.6] Dynamic relations added to Eloquent

### PR DESCRIPTION
**Pull request by kapv89**
Added functionality to add and remove new relations to an Eloquent Model. Can be useful in projects with a core set of models which provide a core set of functionality, and service-module specific models which provide additional, opt-able set of functionality.

For example:

```php
class Post extends Model {
...
}

Post::addDynamicRelation('comments', function ($m) {
    $m->hasMany(Comment::class, 'post_id');
});
...
Post::removeDynamicRelation('comments');

```

In the above written code, assume that the model `Post` is part of the `Core` package, and the model `Comment` is part of `Comments` service-module.

### CODE REVIEW

**Comment by browner12** [comment]
can you provide a more in depth solution of why this is needed?  I'm not connecting the dots why we would need to establish relationships at runtime.

**Comment by kapv89** [comment]
Hey @browner12 

So assume that you are building something like domain-specific-CMS. Now suppose you sell commercial add-ons of this CMS as separate packages which contain service-providers.

When such a service provider is included in the CMS configuration, the service-provider's models will be registered with the application, and they can establish relationships dynamically with the core CMS models.

This is one scenario that I can think of where this feature will be useful. There can be several other scenarios where the project requirement would demand that certain parts of the domain-model be toggled on, or off in the system as and when needed. For all such cases, this feature will be useful.

I hope I was clear.

**Comment by kapv89** [comment]
Can someone please comment on the styleci analysis? I am not sure what I am missing on.

**Comment by browner12** [comment]
if you click the "details" link it will show you exactly what's wrong. you have 3 spaces instead of 4, trailing spaces, and opening curly braces on the wrong line.

**Comment by pavinthan** [comment]
@k1ng440 change the code style please

**Comment by browner12** [comment]
definitely an interesting idea.

does this mean the Service Provider will have to have knowledge of the models that exist in the main app? would we need to build in a check to make sure the model exists?

**Comment by kapv89** [comment]
Yes, the Service Provider will have to have the knowledge of the models in the main app.

Hypothetically, any system where the main app has an architecture that allows plugins to exist for the core domain logic, will require that its plugins know the pieces of the main architecture which they need to interact with.

Coming to building a check which makes sure that the related model exists, I am not sure how it can be done with this particular implementation of dynamic-relations. As in this implementation, the related model is defined inside the closure. Relevant piece of code pasted below:

```php
Post::addDynamicRelation('comments', function ($m) {
    $m->hasMany(Comment::class, 'post_id');
});
```

**Comment by deleugpn** [comment]
When combining a world of micro-services with the simplicity of implementing this feature, it seem a nice-to-have tool in your belt. However, it doesn't really feel core-related feature. I think I would like this better as a separate trait that you include in your model. You could have one general model that `extends` Eloquent Model and include your trait there or on a per-model basis.

**Comment by rs-sliske** [comment]
seems interesting idea, but can this not be achieved with macroable?

**Comment by sisve** [comment]
I don't see why this is a good idea. How would two plugins work with this; how do you avoid that two plugins wouldn't attempt to add relations with identical names? This is bound for conflicts.

Do you realize that `Post::addDynamicRelation('comments', ...);` would add the relation to all your models, including other types like User? The $dynamicRelations array is static on the Model class after all...

**Comment by kapv89** [comment]
@sisve I don't think that the problem you are mentioning will occur because of the `static` works(late static binding and all). However, just to be sure, I'll add a test to verify this

**Comment by sisve** [comment]
My point about duplicate names wasn't that you should add a fix for it, but that the whole idea is broken. Who decides which service provider has the right for a specific relationship name?

Alternatives that already exists, already works and are fully known for all php developers;
1. Plugins provides a opt-in trait called HasComments.
2. Plugins provides a repository/facade. `Comments::getForPost($post)`

How would you add a dynamic relation to all models implementing a Commentable interface? Can you even use an interface to enforce the existence of a comments relation with this solution?

Who owns the models and views in your scenario? Normally models say "I can have comments", or the views says "show the comments for the model here". In your case you would have a service provider that says "the model has comments", but you would still need to update your views. I don't see why we would add all this magic if you need to edit code in your application anyway; why not just use a single line of code using a trait that the plugin provides?

**Comment by kapv89** [comment]
@sisve Found some issues with the `static` thing. Let me deal with that. I'll get to "why this is a good idea" once I push some fixes. Thanks for pointing them out though!

**Comment by kapv89** [comment]
Hi @sisve 

Assume that you are working in a team of 10 on a project which has a `core` module, and 3 other separate modules, `foo`, `bar`, and `notifications`.

Now suppose that the `core` module is the one serving an API over which you have mobile apps, and web apps with which the users interact. This `core` module contains all the UX related logic which enables the user-interactions that happen with your mobile apps, and web apps. It also contains the domain logic which defines what the software you are building is about.

Along with all this, there would be a set of developers working on this `core` module, who can act as knowledge base for other developers working on client-apps, and modules `foo`, `bar`, and `notifications`.

Now let's get to the `notifications` module. This might start out as a simple module, but as the software grows, and the user-base grows, it'll have its own set of intricacies that'll have to be dealt with. For example, let's say a user takes 15 actions in 5 minutes. Now according to business requirements, these 15 actions will have to result in 3 different types of notifications that should be sent out to 2 other users. All this logic is going to reside in the `notifications` module.

Because of the complexity of `notifications` module, and given that the logic related to notifications is not a part of the domain logic that the module `core` is responsible for, the module `notifications` is going to have some of its own Eloquent models. These Eloquent models are never going to be used in the modules `core`, `foo`, or `bar`. However the domain logic of `notifications` module would be easier to write if some models in the module `core` have relations with the models in module `notifications`.

Now that the scenario is set (and I hope that you can see that this is not an uncommon scenario), let's come to the 2 alternatives you pointed out.

I'll start with point number 2 first, i.e., "plugin provides a repository/facade". Assuming that even if the `notifications` module has such a repository, in the scenario when this package needs to generate notifications related to a model from the `core` module(let's call this model `Post` for now), there can be cases where this `notifications` package needs to attach some metadata to the model `Post` (let's represent this metadata by a model `Item` in `notifications` package for now). At times, the developers working on the `notifications` package will need to do something like the following:
```php
$posts = Post::with('notificationItems')->whereIn('fizz', $fizzes)->get();
```
If this is done only with a repository(i.e., without defining a relationship in the `core` module), then the eager-loading logic, which is already provided by Eloquent, will have to be re-written for all the different relations that exists across the boundaries of the modules `core` and `notifications`. Hence, I hope you can see, that point number 2 is not a valid solution for non-trivial scenarios.

Now let's come to point number 1, i.e., "Plugins provides an opt-in trait". This solution comes closer to the problem being solved by dynamic-relationships. This way the module `notifications` can provide a trait which is called `RelatedToNotificationsTrait` which will be used by all the models in the module `core` which are involved in the processes handled by the module `notifications`. So in this scenario, the model `Post` will look like this:
```php
// somewhere inside the module `core`
class Post extends Eloquent\Model {
    use RelatedToNotificationsTrait;
}
```
This presents several problems, all of which are very similar to the problems that will be presented before the feature dynamic-relationships. First, I'll show how the same functionality can be achieved using dynamic relationships. Given below:
```php
// inside the service-provider of the module `notifications`
class NotificationsServiceProvider extends Illuminate\Support\ServiceProvider {
    public function register() {
        Post::addDynamicRelation('notificationMetaItems', ($m) {
            return $m->hasMany(NotificationMetaItem::class, 'post_id');
        });
    }
}
```

Now  I'd compare traits and dynamic-relations and show that the solution offered by dynamic-relations becomes more **developer-friendly** than the one offered by traits with growth in size and complexity of code-base.

**traits** -- In traits, the developers working on the module `core` will have to be aware of the methods provided by the trait `RelatedToNotificationsTrait`(from the module `notifications`) and will have to review the said trait before adding some functionality to the model `Post`. This functionality will mostly be completely unrelated to the module `notifications`, yet a developer working module `core` will have to divert some time in reaching to and going through the trait `RelatedToNotificationsTrait`. Also, while defining this trait, developers working on module `notifications will have to look out for name-conflicts with relations and methods in the model `Post` from the package `core`.

**dynamic-relations** -- In dynamic-relations, a developer working on the module `core` will have to run the code `Post::getAllDynamicRelations()`(via tinker, for example). This will show all the dynamic-relations that have been defined for the model `Post`, and the developer working on the module `core` will have to avoid name-conflicts with the said dynamic-relations shown via the code `Post::getAllDynamicRelations()`. Also, while adding a new dynamic relation to the model `Post`, a developer will have to avoid name conflicts with existing dynamic relations of model `Post`, and with methods defined in the model `Post`.

The solution offered by traits and dynamic-relations will have similar effect on developer productivity for first few modules. However, as the size of the code-base grows, and more modules get added in the system, and more people get involved in the system, dynamic-relationships will become easier to work with when compared to traits.

For example, let's say that are using traits, and over time, the model `Post`(in the module `core`) accumulates 5 different traits from 4 different modules. Adding new functionality to the model `Post` will require checks across 5 different files.

On the other hand, if you are using dynamic-relationships, then the only check that is required before adding a functionality to model `Post` is going through the model `Post`, and running `Post::getAllDynamicRelations()`.

I hope you can see why I am pushing for dynamic-relations. I can present an equally elaborate example with a different type of business-problem which is presented in an equally large system.

PS: made dynamic-relation play well with inheritance in last few pushes.

File: src/Illuminate/Database/Eloquent/Model.php [hunk 1]
```diff
@@ -123,7 +124,7 @@
      *
      * @var array
      */
-    protected static $dynamicRelations = [];
+    public static $dynamicRelations = [];
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Model.php]: Why is this public?

Comment (kapv89) [reply src/Illuminate/Database/Eloquent/Model.php]: Sorry, was trying to make a feature, and kept it public for debugging-related reasons. Will push a fix right away.

Comment (kapv89) [reply src/Illuminate/Database/Eloquent/Model.php]: Fixed!

**Comment by deleugpn** [comment]
All of this can be solved by having a `Package\Notifications\Post extends Core\Post`. Developers working on Core can still provide code for the package teams while the package teams have the flexibility to deal with their own set of particularities.
So, if you're trying to merge this in order to solve the issue you just explained, my personal opinion is that you're doing it wrong and this should not be the way to go.
As for the opt-in that I mentioned before is not about your project code, but rather about this PR; instead of making these changes inside Eloquent Model, you should write a trait for Laravel to offer. Something like `HasDynamicRelationship`. You would then use this trait in your project whenever you want this, meaning not everybody using eloquent will have this by default.

**Comment by kapv89** [comment]
@deleugpn good point, but I still see a problem in the solution that you offered. Sticking with the `core` and `notifications` scenario, `Post` is only one of the `core` models that I had mentioned that can be involved in the processes handled by the `notifications` package.

However if there are around 4 to 5 `core` models involved in the `notifications` module, you'll have that many Eloquent child classes per module level.

Also, I am trying to merge this because I see several use cases for this feature which become helpful as a Laravel project grows beyond a certain size. The issue that I had explained is just one example of scenarios where a feature like dynamic-relations can be helpful.

The first example I had referred to in an earlier comment was that of a "domain specific CMS". I hope you can see how a feature like dynamic-relations can be helpful in a project like this which again has non-trivial complexity.

Moreover, I personally find the feature dynamic-relations to be more inline with the active-record nature of Eloquent

**Comment by sisve** [comment]
> Now let's get to the notifications module. This might start out as a simple module, but as the software grows, and the user-base grows, it'll have its own set of intricacies that'll have to be dealt with. For example, let's say a user takes 15 actions in 5 minutes. Now according to business requirements, these 15 actions will have to result in 3 different types of notifications that should be sent out to 2 other users. All this logic is going to reside in the notifications module.
> 
> Because of the complexity of notifications module, and given that the logic related to notifications is not a part of the domain logic that the module core is responsible for, the module notifications is going to have some of its own Eloquent models. These Eloquent models are never going to be used in the modules core, foo, or bar. However the domain logic of notifications module would be easier to write if some models in the module core have relations with the models in module notifications.

I disagree with the entire application architecture that you describe. Core models should not have hard references to notifications at all; they are two distinct separate domain areas. There's no requirement at all for the notifications plugin to use the same database, or even a database.

You mention eager-loading, but that's just another select-query. It could just as easily be done with a `NotificationsFacade::getForPosts($postIds)` and it would perhaps do a select-query. The important part is that the notifications module is separate, treated as such, and you do not make dependencies from your code to the notifications package.

Imagine that a comments package was backed by Disqus, would you still need Eloquent models? Imagine that a profile-photo package was backed by the filesystem, would you still need Eloquent models? Your solution only makes sense if 1) you are forcing the packages to use Eloquent, 2) you are using the same database. (Imagine all the support issues that will be opened about joins not working if you use separate databases...)

There's nothing developer-friendly about the dynamic relationships; you have to call code at runtime to even figure out which ones are registered. There's no code completion in your ide, no known language features involved, etc.

**Comment by deleugpn** [comment]
1- If you are developing multi-level dependency package like you've described, perhaps rethinking the process of integration between modules could be necessary on your side. The Laravel Framework is unfortunately not the suitable place to discuss that.

2- This is no longer a simple few lines of code that it originally was, so I no longer think it fits Laravel core functionality. I think you can still achieve this as you wish by simply extending Eloquent model to add Dynamic Relationship to it and then have all your application models extend your custom eloquent. If for some reason you think this is not possible, that could be something worth discussing, e.g what is preventing eloquent to be extendable in that sense?

3- Even if this were to get merged, I still think a trait would be a better place to put that code.

4- Even considering the scenario you described (5 models in a 4-level dependency), I still think what you're describing is class extension 101. `Core` provides a basic model and `Notifications` wants to **extend** that model to improve on it. Even if you have another level of dependency that would extend `Package\Notifications\Post`, I still think it's somewhat the better move. It doesn't rely on magic knowledge (like sisve described), the upper level team can still work on their code without previous knowledge of lower level dependencies and the dependents can add on to the models as they need.

**Comment by taylorotwell** [comment]
No plans to add to the core right now.

**Comment by kapv89** [comment]
https://npmjs.com/package/@krab/krab

**Pull request closed without merging**